### PR TITLE
Support multiple engines for non-connection share levels

### PR DESCRIPTION
**Pull request by yaooqinn**
### _Why are the changes needed?_

This PR adds multiple engine support for non-connection share levels. A new configuration called  `kyuubi.engine.share.level.sub.domain` is added for this functionality. For simplification, the configuration only accepts [1, 10]-length case insensitive string, only '_' and alphabets are valid. 

Unlike connection mode, user mode supports the "long-running" feature across different JDBC connections. But the current implementation only is able to create one engine for a user. End-users don't want to put all eggs into one basket. It's not only less robust and flexible for a static engine to handle different workloads, and also hard to leverage failover when there is a zombied engine being cached.

With multiple engines support, end-users are able to
1. create as many engines as they want to meet the concurrency.
2. leverage different queue/namespace resource with long-running mode
3. can give different resources or priorities to their own engines for different workloads
4. can perform fast failover by creating a new sub domain
5. etc.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @turboFei @pan3793 @ulysses-you thanks

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/ShareLevelSparkEngineSuite.scala [hunk 1]
```diff
@@ -33,7 +33,7 @@ abstract class ShareLevelSparkEngineSuite
   extends WithDiscoverySparkSQLEngine with JDBCTestUtils {
   def sharedLevel: ShareLevel
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/ShareLevelSparkEngineSuite.scala:34]: still some `shared` existed

File: docs/deployment/settings.md [hunk 2]
```diff
@@ -219,7 +221,7 @@ kyuubi\.session\.engine<br>\.idle\.timeout|<div style='width: 65pt;word-wrap: br
...
-kyuubi\.session\.engine<br>\.share\.level|<div style='width: 65pt;word-wrap: break-word;white-space: normal'>USER</div>|<div style='width: 170pt;word-wrap: break-word;white-space: normal'>The SQL engine App will be shared in different levels, available configs are: <ul> <li>CONNECTION: the App will not be shared but only used by the current client connection</li> <li>USER: the App will be shared by all sessions created by a unique username</li> <li>SERVER: the App will be shared by Kyuubi servers</li></ul></div>|<div style='width: 30pt'>string</div>|<div style='width: 20pt'>1.0.0</div>
+kyuubi\.session\.engine<br>\.share\.level|<div style='width: 65pt;word-wrap: break-word;white-space: normal'>USER</div>|<div style='width: 170pt;word-wrap: break-word;white-space: normal'>(deprecated) - Using kyuubi.engine.share.level instead</div>|<div style='width: 30pt'>string</div>|<div style='width: 20pt'>1.0.0</div>
```
[the remark this answers is not in the document]
Comment (yaooqinn) [reply docs/deployment/settings.md:224]: maybe the br tag will let the '.' become a leading character

**Comment by yaooqinn** [comment]
thanks, merged to master for v1.2.0
